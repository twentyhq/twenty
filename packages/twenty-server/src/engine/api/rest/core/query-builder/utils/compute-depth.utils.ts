import { BadRequestException } from '@nestjs/common';

import { type Request } from 'express';

const ALLOWED_DEPTH_VALUES = [0, 1, 2];

export const computeDepth = (request: Request): number | undefined => {
  if (!request.query.depth) {
    return undefined;
  }

  const depth = +request.query.depth;

  if (isNaN(depth) || !ALLOWED_DEPTH_VALUES.includes(depth)) {
    throw new BadRequestException(
      `'depth=${
        request.query.depth
      }' parameter invalid. Allowed values are ${ALLOWED_DEPTH_VALUES.join(
        ', ',
      )}`,
    );
  }

  return depth;
};
