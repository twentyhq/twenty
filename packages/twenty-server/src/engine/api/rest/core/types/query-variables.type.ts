import { ObjectRecordOrderBy } from 'src/engine/api/graphql/workspace-query-builder/interfaces/object-record.interface';

export type QueryVariables = {
  id?: string;
  ids?: string[];
  data?: object | null;
  filter?: object;
  orderBy?: ObjectRecordOrderBy;
  last?: number;
  first?: number;
  startingAfter?: string;
  endingBefore?: string;
  input?: object;
};
