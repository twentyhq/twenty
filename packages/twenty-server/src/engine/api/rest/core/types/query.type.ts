import { type QueryVariables } from 'src/engine/api/rest/core/types/query-variables.type';

export type Query = {
  query: string;
  variables: QueryVariables;
};
