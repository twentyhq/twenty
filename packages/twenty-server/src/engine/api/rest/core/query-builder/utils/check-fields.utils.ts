import { BadRequestException } from '@nestjs/common';

import { compositeTypeDefinitions } from 'src/engine/metadata-modules/field-metadata/composite-types';
import { isCompositeFieldMetadataType } from 'src/engine/metadata-modules/field-metadata/utils/is-composite-field-metadata-type.util';
import { ObjectMetadataItemWithFieldMaps } from 'src/engine/metadata-modules/types/object-metadata-item-with-field-maps';
import { computeObjectTargetTable } from 'src/engine/utils/compute-object-target-table.util';

export const checkFields = (
  objectMetadataItem: ObjectMetadataItemWithFieldMaps,
  fieldNames: string[],
): void => {
  const fieldMetadataNames = objectMetadataItem.fields
    .map((field) => {
      if (isCompositeFieldMetadataType(field.type)) {
        const compositeType = compositeTypeDefinitions.get(field.type);

        if (!compositeType) {
          throw new BadRequestException(
            `Composite type '${field.type}' not found`,
          );
        }

        // TODO: Don't really know why we need to put fieldName and compositeType name here
        return [
          field.name,
          compositeType.properties.map(
            (compositeProperty) => compositeProperty.name,
          ),
        ].flat();
      }

      return field.name;
    })
    .flat();

  for (const fieldName of fieldNames) {
    if (!fieldMetadataNames.includes(fieldName)) {
      throw new BadRequestException(
        `field '${fieldName}' does not exist in '${computeObjectTargetTable(
          objectMetadataItem,
        )}' object`,
      );
    }
  }
};
