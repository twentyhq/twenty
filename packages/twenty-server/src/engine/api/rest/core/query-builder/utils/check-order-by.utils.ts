import { BadRequestException } from '@nestjs/common';

import { ObjectRecord } from 'src/engine/api/graphql/workspace-query-builder/interfaces/object-record.interface';

import { compositeTypeDefinitions } from 'src/engine/metadata-modules/field-metadata/composite-types';
import { isCompositeFieldMetadataType } from 'src/engine/metadata-modules/field-metadata/utils/is-composite-field-metadata-type.util';
import { ObjectMetadataItemWithFieldMaps } from 'src/engine/metadata-modules/types/object-metadata-item-with-field-maps';
import { computeObjectTargetTable } from 'src/engine/utils/compute-object-target-table.util';

export const checkArrayFields = (
  objectMetadataItem: ObjectMetadataItemWithFieldMaps,
  fields: Array<Partial<ObjectRecord>>,
): void => {
  const fieldMetadataNames = objectMetadataItem.fields
    .map((field) => {
      if (isCompositeFieldMetadataType(field.type)) {
        const compositeType = compositeTypeDefinitions.get(field.type);

        if (!compositeType) {
          throw new BadRequestException(
            `Composite type '${field.type}' not found`,
          );
        }

        return [
          field.name,
          compositeType.properties.map(
            (compositeProperty) => compositeProperty.name,
          ),
        ].flat();
      }

      return field.name;
    })
    .flat();

  for (const fieldObj of fields) {
    for (const fieldName in fieldObj) {
      if (!fieldMetadataNames.includes(fieldName)) {
        throw new BadRequestException(
          `field '${fieldName}' does not exist in '${computeObjectTargetTable(
            objectMetadataItem,
          )}' object`,
        );
      }
    }
  }
};
