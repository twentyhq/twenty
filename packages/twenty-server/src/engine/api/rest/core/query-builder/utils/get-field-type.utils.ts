import { FieldMetadataType } from 'twenty-shared/types';

import { ObjectMetadataItemWithFieldMaps } from 'src/engine/metadata-modules/types/object-metadata-item-with-field-maps';

export const getFieldType = (
  objectMetadataItem: ObjectMetadataItemWithFieldMaps,
  fieldName: string,
): FieldMetadataType | undefined => {
  return objectMetadataItem.fieldsByName[fieldName]?.type;
};
