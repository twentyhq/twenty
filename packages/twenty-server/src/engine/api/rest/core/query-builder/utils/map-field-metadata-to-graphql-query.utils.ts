import { FieldMetadataType } from 'twenty-shared/types';

import { FieldMetadataInterface } from 'src/engine/metadata-modules/field-metadata/interfaces/field-metadata.interface';

import { RelationMetadataType } from 'src/engine/metadata-modules/relation-metadata/relation-metadata.entity';
import { ObjectMetadataMaps } from 'src/engine/metadata-modules/types/object-metadata-maps';

const DEFAULT_DEPTH_VALUE = 1;

// TODO: Should be properly type and based on composite type definitions
export const mapFieldMetadataToGraphqlQuery = (
  objectMetadataMaps: ObjectMetadataMaps,
  field: FieldMetadataInterface,
  maxDepthForRelations = DEFAULT_DEPTH_VALUE,
): string | undefined => {
  if (maxDepthForRelations < 0) {
    return '';
  }

  const fieldType = field.type;

  const fieldIsSimpleValue = [
    FieldMetadataType.UUID,
    FieldMetadataType.TEXT,
    FieldMetadataType.DATE_TIME,
    FieldMetadataType.DATE,
    FieldMetadataType.BOOLEAN,
    FieldMetadataType.NUMBER,
    FieldMetadataType.NUMERIC,
    FieldMetadataType.RATING,
    FieldMetadataType.SELECT,
    FieldMetadataType.MULTI_SELECT,
    FieldMetadataType.POSITION,
    FieldMetadataType.RAW_JSON,
    FieldMetadataType.RICH_TEXT,
    FieldMetadataType.ARRAY,
    FieldMetadataType.TS_VECTOR,
  ].includes(fieldType);

  if (fieldIsSimpleValue) {
    return field.name;
  } else if (
    maxDepthForRelations > 0 &&
    fieldType === FieldMetadataType.RELATION &&
    field.toRelationMetadata?.relationType === RelationMetadataType.ONE_TO_MANY
  ) {
    const fromObjectMetadataId = field.toRelationMetadata?.fromObjectMetadataId;

    if (!fromObjectMetadataId) {
      return '';
    }

    const relationMetadataItem = objectMetadataMaps.byId[fromObjectMetadataId];

    if (!relationMetadataItem) {
      return '';
    }

    return `${field.name}
    {
      id
      ${Object.values(relationMetadataItem.fieldsById)
        .map((field) =>
          mapFieldMetadataToGraphqlQuery(
            objectMetadataMaps,
            field,
            maxDepthForRelations - 1,
          ),
        )
        .join('\n')}
    }`;
  } else if (
    maxDepthForRelations > 0 &&
    fieldType === FieldMetadataType.RELATION &&
    field.fromRelationMetadata?.relationType ===
      RelationMetadataType.ONE_TO_MANY
  ) {
    const toObjectMetadataId = field.fromRelationMetadata?.toObjectMetadataId;

    if (!toObjectMetadataId) {
      return '';
    }

    const relationMetadataItem = objectMetadataMaps.byId[toObjectMetadataId];

    if (!relationMetadataItem) {
      return '';
    }

    return `${field.name}
      {
        edges {
          node {
            id
            ${Object.values(relationMetadataItem.fieldsById)
              .map((field) =>
                mapFieldMetadataToGraphqlQuery(
                  objectMetadataMaps,
                  field,
                  maxDepthForRelations - 1,
                ),
              )
              .join('\n')}
          }
        }
      }`;
  } else if (fieldType === FieldMetadataType.LINKS) {
    return `
      ${field.name}
      {
        primaryLinkLabel
        primaryLinkUrl
        secondaryLinks
      }
    `;
  } else if (fieldType === FieldMetadataType.CURRENCY) {
    return `
      ${field.name}
      {
        amountMicros
        currencyCode
      }
    `;
  } else if (fieldType === FieldMetadataType.FULL_NAME) {
    return `
      ${field.name}
      {
        firstName
        lastName
      }
    `;
  } else if (fieldType === FieldMetadataType.ADDRESS) {
    return `
      ${field.name}
      {
        addressStreet1
        addressStreet2
        addressCity
        addressPostcode
        addressState
        addressCountry
        addressLat
        addressLng
      }
    `;
  } else if (fieldType === FieldMetadataType.ACTOR) {
    return `
      ${field.name}
      {
        source
        workspaceMemberId
        name
      }
    `;
  } else if (fieldType === FieldMetadataType.EMAILS) {
    return `
      ${field.name}
      {
        primaryEmail
        additionalEmails
      }
    `;
  } else if (fieldType === FieldMetadataType.PHONES) {
    return `
      ${field.name}
      {
        primaryPhoneNumber
        primaryPhoneCountryCode
        primaryPhoneCallingCode
        additionalPhones
      }
    `;
  } else if (fieldType === FieldMetadataType.RICH_TEXT_V2) {
    return `
      ${field.name}
      {
        blocknote
        markdown
      }
    `;
  }
};
