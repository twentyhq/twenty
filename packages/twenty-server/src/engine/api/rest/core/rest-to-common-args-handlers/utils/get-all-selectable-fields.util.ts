import {
  FieldMetadataType,
  RelationType,
  type RestrictedFieldsPermissions,
} from 'twenty-shared/types';
import { isDefined } from 'twenty-shared/utils';

import { compositeTypeDefinitions } from 'src/engine/metadata-modules/field-metadata/composite-types';
import { isCompositeFieldMetadataType } from 'src/engine/metadata-modules/field-metadata/utils/is-composite-field-metadata-type.util';
import { type ObjectMetadataItemWithFieldMaps } from 'src/engine/metadata-modules/types/object-metadata-item-with-field-maps';
import { isFieldMetadataEntityOfType } from 'src/engine/utils/is-field-metadata-of-type.util';

type SelectableFieldsStructured = Record<
  string,
  boolean | Record<string, boolean>
>;

export const getAllSelectableFields = ({
  restrictedFields,
  objectMetadata,
}: {
  restrictedFields: RestrictedFieldsPermissions;
  objectMetadata: { objectMetadataMapItem: ObjectMetadataItemWithFieldMaps };
}): SelectableFieldsStructured => {
  const result: SelectableFieldsStructured = {};

  const fields = Object.values(objectMetadata.objectMetadataMapItem.fieldsById);

  for (const field of fields) {
    if (restrictedFields[field.id]?.canRead === false) continue;

    if (isCompositeFieldMetadataType(field.type)) {
      const compositeType = compositeTypeDefinitions.get(field.type);

      if (!compositeType) {
        throw new Error(
          `Composite type definition not found for type: ${field.type}`,
        );
      }

      const compositeFields: Record<string, boolean> = {};

      for (const property of compositeType.properties) {
        compositeFields[property.name] = true;
      }

      result[field.name] = compositeFields;
    } else if (
      isFieldMetadataEntityOfType(field, FieldMetadataType.RELATION) &&
      field.settings.relationType === RelationType.MANY_TO_ONE &&
      isDefined(field.settings.joinColumnName)
    ) {
      result[field.settings.joinColumnName] = true;
    } else {
      result[field.name] = true;
    }
  }

  return result;
};
