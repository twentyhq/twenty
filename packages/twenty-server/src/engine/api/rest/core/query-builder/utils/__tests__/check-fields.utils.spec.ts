import { FieldMetadataInterface } from 'src/engine/metadata-modules/field-metadata/interfaces/field-metadata.interface';

import {
  fieldNumberMock,
  objectMetadataItemMock,
} from 'src/engine/api/__mocks__/object-metadata-item.mock';
import { checkFields } from 'src/engine/api/rest/core/query-builder/utils/check-fields.utils';
import { checkArrayFields } from 'src/engine/api/rest/core/query-builder/utils/check-order-by.utils';
import { FieldMetadataMap } from 'src/engine/metadata-modules/types/field-metadata-map';

describe('checkFields', () => {
  const completeFieldNumberMock: FieldMetadataInterface = {
    id: 'field-number-id',
    type: fieldNumberMock.type,
    name: fieldNumberMock.name,
    label: 'Field Number',
    objectMetadataId: 'object-metadata-id',
    isNullable: fieldNumberMock.isNullable,
    defaultValue: fieldNumberMock.defaultValue,
  };

  const fieldsById: FieldMetadataMap = {
    'field-number-id': completeFieldNumberMock,
  };

  const fieldsByName: FieldMetadataMap = {
    [completeFieldNumberMock.name]: completeFieldNumberMock,
  };

  const mockObjectMetadataWithFieldMaps = {
    ...objectMetadataItemMock,
    fieldsById,
    fieldsByName,
    fieldsByJoinColumnName: {},
  };

  it('should check field types', () => {
    expect(() =>
      checkFields(mockObjectMetadataWithFieldMaps, ['fieldNumber']),
    ).not.toThrow();

    expect(() =>
      checkFields(mockObjectMetadataWithFieldMaps, ['wrongField']),
    ).toThrow();

    expect(() =>
      checkFields(mockObjectMetadataWithFieldMaps, [
        'fieldNumber',
        'wrongField',
      ]),
    ).toThrow();
  });

  it('should check field types from array of fields', () => {
    expect(() =>
      checkArrayFields(mockObjectMetadataWithFieldMaps, [
        { fieldNumber: undefined },
      ]),
    ).not.toThrow();

    expect(() =>
      checkArrayFields(mockObjectMetadataWithFieldMaps, [
        { wrongField: undefined },
      ]),
    ).toThrow();

    expect(() =>
      checkArrayFields(mockObjectMetadataWithFieldMaps, [
        { fieldNumber: undefined },
        { wrongField: undefined },
      ]),
    ).toThrow();
  });
});
