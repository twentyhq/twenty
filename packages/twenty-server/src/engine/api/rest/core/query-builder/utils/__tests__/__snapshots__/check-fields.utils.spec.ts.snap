// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checkFields should accept valid relation field name 1`] = `"field 'pointOfContact' does not exist in 'opportunity' object"`;

exports[`checkFields should reject invalid field name 1`] = `"field 'wrongField' does not exist in 'opportunity' object"`;

exports[`checkFields should reject mix of valid and invalid field names 1`] = `"field 'wrongField' does not exist in 'opportunity' object"`;
